Correction validation

QuantumEF opened the pull request:
Improvements to validate the values used to create "corrections" since while their type is i16, they can only range between certain values.

QuantumEF commented:
Realized I still need to add tests as well as investigate to see if I can remove some code duplication.

QuantumEF commented:
How do you feel about the Add implementations?

TroyNeubauer commented:
I think exposing a `saturating_add` totally makes sense. Its a bit annoying that `num_traits::ops::saturating::SaturatingAdd` has `Add` as a supertrait since I think its a bit confusing to have `Add` implicitly saturate

This approach makes the most sense to me:
1. Have `fn new(i16) -> Option<Self>` [DONE]
2. Add a `fn new_saturating(i16) -> Self` which takes any value returns a saturated wrapper type
3. Have a freestanding `fn saturating_add(&self, Self) -> Self` method

That way users have decent control / introspection into the values without `Add` implicitly saturating. For more complicated logic, I think it would be easier for them to convert back to `i16`s, do whatever computation, then saturate back into the allowed range or fail if it falls out of range. 

What do you think?

TroyNeubauer commented:
Oh I see. The rust approach for the built in types (`u32`, `usize`, etc.)  is to panic in debug builds and wrap "soft UB" in release.

For these, since the range is a lot more constrained, things like wrapping, saturating, etc. is more likely to happen and for that reason I think the APIs should be explicit.

What do you think about constraining these types even further (so that the user has full control over what goes on when doing computations with `i16`), and having them only convert to the wrapper type at the end?

I.E minify the API to be just:
1. `fn new(i16) -> Option<Self>`
2. `fn new_saturating(i16) -> Self` which takes any value and returns a saturated wrapper type
3. `Self::MIN` + `Self::MAX`, so they can check their work along the way

That way we avoid having to make decisions about what the user wants. We give them the minimal set of tools, and have them decide how to handle needing a gain / bias that is out of range (maybe their hardware is bad, or a computation showing they need a gain higher than the radio supports means the DC compensation will be less effective!). I think making that explicit with a smaller API is a good idea.

QuantumEF commented:
I think that seems reasonable.

QuantumEF commented:
I removed those traits and added the `new_saturating()` method as well as `min()` and `max()` methods that return `Self(MIN)` and `Self(MAX)` respectively.

TroyNeubauer commented on src/bladerf.rs:
```diff
@@ -843,7 +840,8 @@ pub trait BladeRF: Sized + Drop {
         check_res!(res);
-        Ok(value)
+        // Safety: the bladerf should return a valid value in the correct range.
+        Ok(unsafe { T::new_unchecked(value) })
```
I would rather panic here if something goes wrong with the bladerf than trigger UB. Looks like the inner call locks a mutex and does some IO so the performance cost of checking should be negligible

QuantumEF replied:
I forgot about that. I will just return some `Err()` since it is already returning a result.

QuantumEF replied:
This does remind me, I created [this issue](https://github.com/Nuand/bladeRF/issues/997) recently.

Does it occur on your device? (I might need to update the program I was using to test that out for you, otherwise you can just use that C program I linked.)

TroyNeubauer reviewed (approved):
One nit but looks good overall!
